### PR TITLE
Add `convert_lit_checkpoint::check_conversion_supported`

### PR DESCRIPTION
**Pull request by JustinGoheen**
Raises `Exception` to inform users that converting models finetuned with `lora`, `adapter`, and `adapter_v2` is not yet supported.

See #374 and #355.

### CODE REVIEW

File: scripts/convert_lit_checkpoint.py [hunk 1]
```diff
@@ -196,6 +196,20 @@ def maybe_unwrap_state_dict(lit_weights: Dict[str, torch.Tensor]) -> Dict[str, t
     return lit_weights.get("model", lit_weights)
 
 
+def maybe_raise_finetune_warning(lit_weights: Dict[str, torch.Tensor], checkpoint_dir: Path) -> None:
+    weight_names = set([wk.split(".")[-1] for wk in lit_weights.keys()])
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:200]:
```suggestion
    weight_names = {wk.split(".")[-1] for wk in lit_weights}
```

File: scripts/convert_lit_checkpoint.py [hunk 2]
```diff
@@ -196,6 +196,20 @@ def maybe_unwrap_state_dict(lit_weights: Dict[str, torch.Tensor]) -> Dict[str, t
...
+    # adapter
+    if "gating_factor" in weight_names:
+        raise Exception(f"Converting models finetuned with adapter not yet support")
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:203]:
```suggestion
        raise NotImplementedError(f"Converting models finetuned with adapter is not supported")
```

File: scripts/convert_lit_checkpoint.py [hunk 3]
```diff
@@ -196,6 +196,20 @@ def maybe_unwrap_state_dict(lit_weights: Dict[str, torch.Tensor]) -> Dict[str, t
...
+    # adapter v2
+    elif "adapter_bias" in weight_names:
+        raise Exception(f"Converting models finetuned with adapter_v2 not yet support")
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:206]:
```suggestion
        raise NotImplementedError(f"Converting models finetuned with adapter v2 is not supported")
```

File: scripts/convert_lit_checkpoint.py [hunk 4]
```diff
@@ -196,6 +196,20 @@ def maybe_unwrap_state_dict(lit_weights: Dict[str, torch.Tensor]) -> Dict[str, t
...
+    # LoRA or QLoRA
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:207]: LoRA or QLoRA can be supported as long as the adapter weights have been merged into the checkpoint. The error message should describe this

File: tests/test_convert_lit_checkpoint.py [hunk 5]
```diff
@@ -191,3 +191,72 @@ def test_maybe_unwrap_state_dict(tmp_path):
...
+    from finetune.lora import save_lora_checkpoint
+    from scripts.convert_lit_checkpoint import convert_lit_checkpoint
+
+    # fabric is needed for finetune.full::save_checkpoint
+    fabric = L.Fabric(devices=1)
+
+    model_name = "Llama-2-7b-hf"
+    ours_config = Config.from_name(model_name, block_size=8, n_layer=2, n_embd=32, n_head=2, padding_multiple=128)
+    ours_model = GPT(ours_config)
+    ours_model.lora_A = torch.nn.Parameter()
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:207]: Why do you need this? You can enable lora by setting `r` in the config

File: tests/test_convert_lit_checkpoint.py [hunk 6]
```diff
@@ -191,3 +191,72 @@ def test_maybe_unwrap_state_dict(tmp_path):
...
+    # save checkpoint to avoid RunTimeError for PytorchStreamReader
+    save_adapter_checkpoint(fabric, ours_model, ckpt_path)
+
+    with pytest.raises(Exception, match="Converting models finetuned with adapter *"):
+        convert_lit_checkpoint(checkpoint_name=ckpt_path.name, checkpoint_dir=ckpt_path.parent, model_name=model_name)
+
+
+def test_maybe_raise_finetune_warning_adapter_v2(tmp_path):
+    # same config as adapter is required
+    # lit_gpt.config.Config will throw AttributeError for no 'adapter_start_layer'
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:242]: This is expected. Adapter uses its own adapter config

File: tests/test_convert_lit_checkpoint.py (hunk 5 above)
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: implemented in [d16241b](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/d16241bda4a44da3135a3f63d7b3c9f116a08494) by setting `r` and `to_mlp`

File: scripts/convert_lit_checkpoint.py (hunk 3 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: implemented in [d16241b](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/d16241bda4a44da3135a3f63d7b3c9f116a08494)

File: scripts/convert_lit_checkpoint.py (hunk 2 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: implemented in [d16241b](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/d16241bda4a44da3135a3f63d7b3c9f116a08494)

File: scripts/convert_lit_checkpoint.py (hunk 4 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py:207]: Updated in [1d54da0](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/1d54da0465d7d2e97c8d17ef770ff31c1c8220ce)

File: scripts/convert_lit_checkpoint.py [hunk 7]
```diff
@@ -196,6 +196,19 @@ def maybe_unwrap_state_dict(lit_weights: Dict[str, torch.Tensor]) -> Dict[str, t
     return lit_weights.get("model", lit_weights)
 
 
+def maybe_raise_finetune_warning(lit_weights: Dict[str, torch.Tensor]) -> None:
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:199]: This says warning but it's an exception, I suggest a different name:

```suggestion
def check_conversion_supported(lit_weights: Dict[str, torch.Tensor]) -> None:
```

File: tests/test_convert_lit_checkpoint.py [hunk 8]
```diff
@@ -191,3 +191,79 @@ def test_maybe_unwrap_state_dict(tmp_path):
     with mock.patch("scripts.convert_lit_checkpoint.maybe_unwrap_state_dict") as maybe_unwrap:
         convert_lit_checkpoint(checkpoint_name=ckpt_name, out_dir=tmp_path, model_name=model_name)
     maybe_unwrap.assert_called()
+
+
+def test_maybe_raise_finetune_warning_adapter(tmp_path):
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:196]: You'll need to update these test names too

File: scripts/convert_lit_checkpoint.py [hunk 9]
```diff
@@ -196,6 +196,19 @@ def maybe_unwrap_state_dict(lit_weights: Dict[str, torch.Tensor]) -> Dict[str, t
...
+    weight_names = {wk.split(".")[-1] for wk in lit_weights}
+    # adapter
+    if "gating_factor" in weight_names:
+        raise NotImplementedError(f"Converting models finetuned with adapter not yet supported.")
+    # adapter v2
+    elif "adapter_bias" in weight_names:
+        raise NotImplementedError(f"Converting models finetuned with adapter_v2 not yet supported.")
+    # LoRA or QLoRA
+    elif any(["lora" in wn for wn in weight_names]):
```
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: updated in [eeb5d2d](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/eeb5d2d91df05c1282d0703ff2c92cc7442e0814)

File: tests/test_convert_lit_checkpoint.py [hunk 10]
```diff
@@ -191,3 +191,79 @@ def test_maybe_unwrap_state_dict(tmp_path):
...
+    from finetune.adapter_v2 import save_adapter_v2_checkpoint
+    from scripts.convert_lit_checkpoint import convert_lit_checkpoint
+
+    # fabric is needed for finetune.full::save_checkpoint
+    fabric = L.Fabric(devices=1)
+
+    model_name = "Llama-2-7b-hf"
+    ours_config = Config.from_name(model_name, block_size=8, n_layer=2, n_embd=32, n_head=2, padding_multiple=128)
+    ours_model = GPT(ours_config)
+    ours_model.adapter_bias = torch.nn.Parameter()
```
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: implemented in [eeb5d2d](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/eeb5d2d91df05c1282d0703ff2c92cc7442e0814)

File: tests/test_convert_lit_checkpoint.py [hunk 11]
```diff
@@ -191,3 +191,79 @@ def test_maybe_unwrap_state_dict(tmp_path):
...
+    from finetune.adapter import save_adapter_checkpoint
+    from scripts.convert_lit_checkpoint import convert_lit_checkpoint
+
+    # fabric is needed for finetune.full::save_checkpoint
+    fabric = L.Fabric(devices=1)
+
+    model_name = "Llama-2-7b-hf"
+    ours_config = Config.from_name(model_name, block_size=8, n_layer=2, n_embd=32, n_head=2, padding_multiple=128)
+    ours_model = GPT(ours_config)
+    ours_model.gating_factor = torch.nn.Parameter()
```
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: implemented in [5d6911a](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/5d6911a1cf3b59ca7d24ac3b9e455c936604dc69)

File: tests/test_convert_lit_checkpoint.py [hunk 12]
```diff
@@ -191,3 +191,79 @@ def test_maybe_unwrap_state_dict(tmp_path):
...
+        r=1,
+        to_mlp=True,
+        block_size=8,
+        n_layer=2,
+        n_embd=32,
+        n_head=2,
+        padding_multiple=128,
+    )
+    ours_model = GPT(ours_config)
+    print(ours_model.state_dict().keys())
```
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: removed in [eeb5d2d](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/eeb5d2d91df05c1282d0703ff2c92cc7442e0814)

File: tests/test_convert_lit_checkpoint.py
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py:1]: implemented in [3638666](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/3638666d91b1ee6d00306212f09a9d37a658cdc6)

File: tests/test_convert_lit_checkpoint.py (hunk 8 above)
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: implemented in [3638666](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/3638666d91b1ee6d00306212f09a9d37a658cdc6)

File: tests/test_convert_lit_checkpoint.py [hunk 13]
```diff
@@ -191,3 +191,79 @@ def test_maybe_unwrap_state_dict(tmp_path):
...
+def test_maybe_raise_finetune_warning_adapter(tmp_path):
+    from lit_gpt.adapter import GPT, Config
+    from finetune.adapter import save_adapter_checkpoint
+    from scripts.convert_lit_checkpoint import convert_lit_checkpoint
+
+    # fabric is needed for finetune.full::save_checkpoint
+    fabric = L.Fabric(devices=1)
+
+    model_name = "Llama-2-7b-hf"
+    ours_config = Config.from_name(model_name, block_size=8, n_layer=2, n_embd=32, n_head=2, padding_multiple=128)
```
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: implemented in [3638666](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/3638666d91b1ee6d00306212f09a9d37a658cdc6)

File: scripts/convert_lit_checkpoint.py [hunk 14]
```diff
@@ -196,6 +196,19 @@ def maybe_unwrap_state_dict(lit_weights: Dict[str, torch.Tensor]) -> Dict[str, t
     return lit_weights.get("model", lit_weights)
 
 
+def check_conversion_supported(lit_weights: Dict[str, torch.Tensor]) -> None:
+    weight_names = {wk.split(".")[-1] for wk in lit_weights}
+    # LoRA or QLoRA
+    if any("lora" in wn for wn in weight_names):
+        raise NotImplementedError(f"Model weights must be merged using `lora.merge_lora_weights()` before conversion.")
```
Comment (JustinGoheen) [inline scripts/convert_lit_checkpoint.py:203]: @carmocca would you prefer for this to be an Exception instead of NotImplementedError?

Comment (carmocca) [reply scripts/convert_lit_checkpoint.py:203]: Maybe ValueError

File: tests/test_convert_lit_checkpoint.py [hunk 15]
```diff
@@ -191,3 +191,30 @@ def test_maybe_unwrap_state_dict(tmp_path):
     with mock.patch("scripts.convert_lit_checkpoint.maybe_unwrap_state_dict") as maybe_unwrap:
         convert_lit_checkpoint(checkpoint_name=ckpt_name, out_dir=tmp_path, model_name=model_name)
     maybe_unwrap.assert_called()
+
+
+def test_check_conversion_supported_adapter(tmp_path):
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:196]: Tmp_path is unused across tests

```suggestion
def test_check_conversion_supported_adapter():
```

**Review by carmocca — APPROVED** [review]
LGTM after the pending comments!

File: tests/test_convert_lit_checkpoint.py (hunk 15 above)
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: updated in [9ab402f](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/9ab402f44120917cd6f0f4b049fcad1be62be24d)

File: scripts/convert_lit_checkpoint.py (hunk 14 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: implemented in [9ab402f](https://github.com/Lightning-AI/lit-gpt/pull/397/commits/9ab402f44120917cd6f0f4b049fcad1be62be24d)

**Pull request merged**